drivers/console/ipm_console.c

```diff
@@ -11,12 +11,62 @@
 #include <logging/log.h>
 LOG_MODULE_REGISTER(ipm_console, CONFIG_IPM_LOG_LEVEL);
 
-static struct device *ipm_console_device;
+#if defined(CONFIG_IPM_CONSOLE_IMMEDIATE)
+struct device *ipm_dev;
+#else
+/*
+ * Stack for the message queue handling thread
+ */
+static K_THREAD_STACK_DEFINE(thread_stack, 1024);
+static struct k_thread thread_data;
+
+K_MSGQ_DEFINE(msg_queue, sizeof(char), CONFIG_IPM_CONSOLE_QUEUE_SIZE, 4);
+
+static void msgq_thread(void *arg1, void *arg2, void *arg3)
+{
+	struct device *ipm_dev = arg1;
+
+	ARG_UNUSED(arg2);
+	ARG_UNUSED(arg3);
+
+	LOG_DBG("started thread");
+
+	while (true) {
+		int character = 0;
+		int ret;
+
+		/* Get character from message queue */
+		k_msgq_get(&msg_queue, &character, K_FOREVER);
+
+		/* Blocking write over IPM */
+		ret = ipm_send(ipm_dev, 1, character, NULL, 0);
+		if (ret) {
+			LOG_ERR("Error sending character %c over IPM, ret %d",
+				character, ret);
+		}
+
+		k_yield();
+	}
+}
+#endif /* ! CONFIG_IPM_CONSOLE_IMMEDIATE */
 
 static int console_out(int c)
 {
-	/* Send character with busy wait */
-	ipm_send(ipm_console_device, 1, c & ~BIT(31), NULL, 0);
+	int ret;
+
+#if defined(CONFIG_IPM_CONSOLE_IMMEDIATE)
+	ret = ipm_send(ipm_dev, 1, c & ~BIT(31), NULL, 0);
+	if (ret) {
+		LOG_ERR("Error sending character %c over IPM, ret %d", c, ret);
+	}
+#else
+	char character = c;
+
+	ret = k_msgq_put(&msg_queue, &character, K_NO_WAIT);
```

Comment:
What's the design goal here?  Why are we implementing asynchronous console output and for who?  The older design was a simple ring buffer which was async by design without the need for complicated and expensive IPC on the device.

Basically this driver creeps me out a bit.  What's the point?